naga/xtask/src/validate.rs

```diff
@@ -13,15 +13,28 @@ use crate::{
     process::{which, EasyCommand},
 };
 
+type Job = Box<dyn FnOnce() -> anyhow::Result<()> + Send + 'static>;
+
 pub(crate) fn validate(cmd: ValidateSubcommand) -> anyhow::Result<()> {
     let mut jobs = vec![];
     collect_validation_jobs(&mut jobs, cmd)?;
 
     let progress_bar = indicatif::ProgressBar::new(jobs.len() as u64);
 
+    let (tx_results, rx_results) = std::sync::mpsc::channel();
+    let enqueuing_thread = std::thread::spawn(move || -> anyhow::Result<()> {
+        for job in jobs {
+            let tx_results = tx_results.clone();
+            crate::jobserver::start_job_thread(move || {
+                tx_results.send(job()).unwrap();
+            })?;
+        }
+        Ok(())
+    });
+
     let mut all_good = true;
-    for job in jobs {
-        if let Err(error) = job() {
+    for result in rx_results {
```

Comment:
issue: We don't maintain a strong expectation of "sent X jobs, must get X results back" with this `Iterator` API usage over a single channel. We also make no attempt to `join` on threads that get spun up. Because of this, I'm concerned that this code will miss gaps in the result pipeline that may be introduced by panicking threads. This would be _weird_, admittedly, but it would also a weak link in `xtask`'s ability to report an issue.[^1]

Two directions I can see from here:

1. Accept this limitation, assume things won't break, and understand that we will be handling panicking issues in job threads reactively. AFAICT, we could only introduce problems here _per validation backend_ (which is a small factor for us), and we update our logic (once working) infrequently. It still makes me nervous with the potential for bug infiltration.
2. Iterate on this code in a separate PR, and do something like create a sender and receiver pair per job we submit.[^2] This way, we can force ourselves to acknowledge missing values as errors, too.

[^1]: Given the scale of testing that we're operating at, I don't think that manual inspection of `xtask`'s `stderr` output (which AIUI would record an issue) is likely to happen without a clear signal from CI or the community that something is wrong in the first place.
[^2]: We might pivot to a single-value channel API like [`postage::oneshot`](https://docs.rs/postage/0.5.0/postage/oneshot/index.html), to keep things cleaner.

Reply:
Capturing panics is a legitimate concern.

Reply:
The new version uses `catch_unwind` while running the job, leaving the `unwrap` after the channel send outside the `catch`, which seems correct. Let me know what you think.

Reply:
LGTM!